rust/src/writer.rs

```diff
@@ -0,0 +1,254 @@
+            txns: vec![],
+        })
+    }
+
+    /**
+     * Return the total Values pending in the buffer
+     */
+    pub fn count(&self, partitions: &WriterPartition) -> Option<usize> {
+        if let Some(buffer) = self.buffer.get(&partitions) {
+            return Some(buffer.len());
+        }
+        None
+    }
+
+    /**
+     * Add a txn action to the buffer
+     */
+    pub fn record_txn(&mut self, txn: Txn) {
+        self.txns.push(txn);
+    }
+
+    /**
+     * Write a new Value into the buffer
+     */
+    pub fn write(
+        &mut self,
+        value: Value,
+        partitions: WriterPartition,
+    ) -> Result<(), DeltaTableError> {
+        match partitions {
+            WriterPartition::NoPartitions => {
+                if self.partitions.len() > 0 {
+                    return Err(DeltaTableError::SchemaMismatch {
+                        msg: "Table has partitions but noone were supplied on write".to_string(),
+                    });
+                }
+            }
+            WriterPartition::KeyValues { .. } => {
+                if self.partitions.len() == 0 {
+                    return Err(DeltaTableError::SchemaMismatch {
+                        msg: "Table has no partitions yet they were supplied on write".to_string(),
+                    });
+                }
+            }
+        }
+
+        if let Some(buffer) = self.buffer.get_mut(&partitions) {
+            buffer.push(value);
+        } else {
+            self.buffer.insert(partitions, vec![value]);
+        }
+        Ok(())
+    }
+
+    /**
+     * Flush the buffer, causing a write of parquet files for each set of partitioned information
+     * as well as any buffered txn actions
+     *
+     * This will create a single transaction in the delta transaction log
+     */
+    pub async fn flush(&mut self) -> Result<(), DeltaTransactionError> {
+        use arrow::json::reader::Decoder;
+
+        let mut parquet_bufs = vec![];
+
+        for (partitions, values) in self.buffer.iter() {
+            let count = self.count(&partitions).unwrap_or(0);
+            let mut value_iter = InMemValueIter::from_vec(&values);
+            let decoder = Decoder::new(self.schema.clone(), count, None);
+
+            let record_batch = decoder
+                .next_batch(&mut value_iter)
+                .map_err(|source| DeltaTableError::ArrowError { source })?;
+
+            if record_batch.is_none() {
+                return Ok(());
+            }
+
+            let mut pb = ParquetBuffer::try_new(self.schema.clone())?;
+            pb.write_batch(&record_batch.unwrap())?;
+            let _metadata = pb.close();
+            parquet_bufs.push(pb.data());
+        }
+
+        let mut dtx = self.table.create_transaction(None);
+        dtx.add_actions(
+            self.txns
+                .drain(0..)
+                .map(|t| crate::action::Action::txn(t))
+                .collect(),
+        );
+
+        dtx.commit(None).await?;
```

Comment:
we need to add file actions before commit as well?

Reply:
:rofl: I missed the most important part, `add_file`